src/parse/parse_step.go

```diff
@@ -83,39 +83,49 @@ func parse(state *core.BuildState, label, dependent core.BuildLabel, mode core.P
 		return err
 	}
 	state.LogParseResult(label, core.PackageParsed, "Parsed package")
-	// The target likely got activated already, however we activate here to handle psudotargets (:all), and to let this
-	// error when the target doesn't exist.
 
+	// The target likely got activated already, however we activate here to handle pseudo-targets (:all), and to let
+	// this error when the target doesn't exist.
 	return state.ActivateTarget(pkg, label, dependent, mode)
 }
 
-// checkSubrepo checks whether this guy exists within a subrepo. If so we will need to make sure that's available first.
+func inSamePackage(label, dependent core.BuildLabel) bool {
+	return label.Subrepo == dependent.Subrepo && label.PackageName == dependent.PackageName
+}
+
+// checkSubrepo checks if the label we're parsing is within a subrepo, returning that subrepo, if present in the label.
+//
+// The subrepo target can be inferred from the subrepo name using convention i.e. ///foo/bar//:baz has a subrepo label
+// //foo:bar. checkSubrepo parses package foo, expecting a call to `subrepo()` that registers a subrepo named foo/bar,
+// so it can return it.
 func checkSubrepo(state *core.BuildState, label, dependent core.BuildLabel, mode core.ParseMode) (*core.Subrepo, error) {
 	if label.Subrepo == "" {
 		return nil, nil
-	} else if subrepo := state.Graph.Subrepo(label.Subrepo); subrepo != nil {
-		return subrepo, nil
 	}
 
-	sl := label.SubrepoLabel(state)
-
-	// Local subincludes are when we subinclude from a subrepo defined in the current package
-	localSubinclude := label.Subrepo == dependent.Subrepo && label.PackageName == dependent.PackageName && mode.IsForSubinclude()
```

Comment:
This check is slightly different now. We no longer check the parse mode but if my reasoning is right, we shouldn't check the parse mode. If we were to continue under any circumstance, we would lock up because we're parsing the same package.